### ISSUE TITLE
Please move kubeadm to its own repo

### ISSUE THREAD

**Issue by bgrant0607** [kubernetes/kubernetes#35314]
kubeadm is new and nothing else in kubernetes depends on it. It should have gone into a separate repository.  #24343 

We need to figure out how to build releases from multiple repos, anyway:
https://github.com/kubernetes/release/issues/3

I understand we wanted to move quickly and this was done for expedience, but productivity in the kubernetes repo is a zero sum game.

cc @jbeda @mikedanese @roberthbailey @lukemarsden @kubernetes/sig-cluster-lifecycle

**Comment by jbeda** [comment]
This is a great thing to consider once the release issue is fixed.

**Comment by luxas** [comment]
Actually I proposed this already earlier in the process, but it didn't recieve much attention.

I'm a really big +1 on this, as now I'm trying to maintain all the issues/PRs we have, label them correctly, find them, prioritize them, add lgtm for those who haven't got write access, etc etc, and that's quite painful with constantly 4500+ issues and 500+ PRs, the bot that is auto-labeling everything wrong all the time and Github's limited ACL features.

And the Submit Queue is really slowing us down IMO, iteration speed has decreased significantly since the initial contribution we made.

It would be helpful is to be able to set kubeadm-specific milestones and labels.
And it would be much easier to give write access to more people if we had kubeadm in a separate kubernetes/ repo.

**Comment by mikedanese** [comment]
I agree with @jbeda that the release work blocks this. We also need to make sure the apigroup machinary works well out of repo as well.

**Comment by luxas** [comment]
Yes, there are work to be done in a lot of different areas for this to work, but once we've done it, splitting kubectl or something else will be much easier :)

**Comment by bogdando** [comment]
The kubeadm depends on the /usr/local/bin/kube-discovery https://github.com/kubernetes/kubernetes/blob/73e89ba0ece903e82b7005be06465753a4e065f7/cmd/kubeadm/app/master/discovery.go#L74. Shall we move the latter to another repo as well? Because dependencies like that might as well speed down kubeadm devs and block them "to move quickly" and maintain "productivity in the kubernetes repo is a zero sum game" that makes them happy.

So it seems to me that any related code only benefits from living in the main repo with the main code it relates to, ensuring high code standards to be kept. And I can see only downsides if it's separated.

**Comment by dgoodwin** [comment]
FWIW the kubeadm -> kube-discovery dependency is soft, and and likely going away very soon. In the link above it's just referencing an image where kube-discovery is published and the binary within it, it can be built and published separately.

**Comment by luxas** [comment]
Yes, we're not planning to keep kube-discovery for long, and even if we did that, we can have it wherever we want to, since the only usage of it is from the pushed image to `gcr.io`.

**Comment by errordeveloper** [comment]
Also renaming to kube-boot, as suggested in another thread, would make a
lot of sense.

**Comment by luxas** [comment]
@errordeveloper Do you mean kube-discovery should be renamed to kube-boot or should kubeadm be renamed?

**Comment by errordeveloper** [comment]
This would be really great. We, kubeadm contributors, had a lot of pain
recently not being able to move as quickly as we could, e.g. we hit test
infra flakes without having any test that concern kubeadm code as such,
also noise in the issues and PRs had been a problem. Additionally, admin
access for some of kubeadm contributors would help the current situation,
but it seems like a big question for contributors to ask and for existing
admins to answer.

There are multiple ways we could solve build and release questions, we
should probably pick one and see how well it works. Perhaps splitting up
build tools and publishing a build tools container would be a good start.

**Comment by bgrant0607** [comment]
@mikedanese Re. API machinery: The conversion machinery (for configuration?) will be moved to its own repo: #2742. The client library is already in another repo.

**Comment by bgrant0607** [comment]
@luxas The kube-boot suggestion came from here: https://github.com/kubernetes/kubernetes/issues/35226#issuecomment-255216857

**Comment by bgrant0607** [comment]
As for kube-discovery: Anything that is deployed as an image should definitely move to a separate repo. We've already started to do that, such as with the node-problem-detector.

**Comment by jbeda** [comment]
Agree -- I'm hoping we we can delete it entirely though as soon as #30707 stuff gets implemented.

**Comment by luxas** [comment]
I have written down some concrete steps how to do this as well as out objectives for v1.6 here:
https://docs.google.com/document/d/1v4k7eN-eetHbbpTAG7WuioRAMsK_FVGi1yC3VIn4Inw/edit#

TL;DR; As soon as kubeadm uses client-go and all pending PRs in this repo are merged, we should move it over

Also I have some WIP of it here: https://github.com/luxas/kubeadm-source (very rough atm)

So @mikedanese @bgrant0607 @roberthbailey can one of you just create the kubernetes/kubeadm repo already now so we just can start filing and keeping track of issues?
That would be great, thanks!

**Comment by jbeda** [comment]
@luxas -- I'd like us to wait to move it out until there is a story on doing releases across multiple repos.  Right now we are releasing `kubeadm` out of band but we really shouldn't be doing that.  We should, over time, sync up with the main kubernetes release cycle.

**Comment by bgrant0607** [comment]
Thanks, @luxas.

@jbeda At the dev summit, I asked for a list of what specific issues were blocking this (e.g., client-go, API machinery for configuration), and suggested that it would be easier to manage kubeadm issues if they were filed in a separate repo, even before the code is moved, which is a pattern that Github recommends.

**Comment by luxas** [comment]
@bgrant0607 Please create a repo with only a markdown file or something soon so we can get going :+1: 

@jbeda _How_ would you like the releases to actually work?
I guess you want them in the release tarball, but should they be in all CI builds as well?
I do not think so.

Also, should we for example have a `release-1.5` branch in the kubeadm repo that the releasing tools will use when doing a `v1.5.x` release for example and so on... I.e. should  we have the same kinds of important branches in both repos to make it possible to have a "stable" version of kubeadm for a given release there?

We should think about how we want it to work first, I think the code that does that will be really straightforward after all...

**Comment by bogdando** [comment]
@jbeda shall we unblock the patch to add kubeadm to the kube then? We could always remove or deprecate later, but it still can be useful while this issue remains on hold.

**Comment by jbeda** [comment]
@luxas -- we should answer these questions for for other potential broken out binaries -- including `kubectl`.  I think we need to have coordinated release branches, probably some integration testing and coordinated versioning.  The result is that we can package matched binaries into OS pkgs/container images/tarballs.

@bogdando -- I'm assuming that you mean `hyperkube` instead of `kube`.  I'd rather not there.  Right now the general wisdom (probably not recorded anywhere 😦 ) is that we'll have a `kube-master`, `kube-node` and `kubectl` as the binaries we build.  The master/node versions will be variations on `hyperkube`.  However, I could see us merging `kubeadm` in to `kubectl` at some point if we can get everyone comfortable with it.  Or use whatever extension mechanisms will be built for `kubectl` (git style?)

**Comment by bogdando** [comment]
@jbdeda got it, thanks! merging kubeadm into kubectl looks like a very good idea to me, even tho a longterm.

**Comment by mikedanese** [comment]
kubectl is a client binary, but kubeadm is a server binary. You would almost never use kubeadm where you use kubectl and vice versa.

**Comment by luxas** [comment]
Totally agree with @mikedanese 

@jbeda I asked because I've thought about the releasing issue, for example the commits for different repos would definitely not match each other. Right now we're coupling the binaries together in a tarball per commit in the main repo. We need to think about how we should workaround that. (For example by ordering the tarballs/folders in `kubernetes-release-dev` by time and under each folder include the different commits for all relevant kubernetes projects)

e.g.

kubernetes-release-dev/
 - v1.6.0-alpha.0-221116-185215/
    - kubernetes-v1.6.0-alpha.0-255+abcdef0123456/
      - bin/${OS}/${ARCH}/${BINARY}
    - kubeadm-v1.6.0-alpha.0-12+abcasbs1544655/
      - bin/${OS}/${ARCH}/kubeadm
    - hyperkube-v1.6.0-alpha.0-188+aasdbasbdb494/
      - bin/${OS}/${ARCH}/hyperkube

Also, re hyperkube split out: We should probably do that, but we have to figure out how and how often we should revendor the dependencies of `hyperkube` (or just `kube`). But that's another thing.

cc @kubernetes/sig-testing ^

**Comment by jbeda** [comment]
Wrt `kubeadm` being a server binary -- I'm not so sure it is clear cut.  As we do things like bootstrap token management, that is a "client" type thing.  Similarly, as we streamline the TLS bootstrap stuff and break that out into separate "phase" commands these things may overlap with `kubectl`.

In addition, it isn't a "server" in that it isn't a daemon process but rather a short lived "client" in the sense that it is used to manipulate other things, sometimes remotely.

But this is mostly moot as we wouldn't merge it for a bit regardless, I think.

**Comment by bgrant0607** [comment]
With respect to release bundling:

If I understood @luxas correctly, it sounds like kubeadm from the K8s release bundle isn't even used.

**Comment by mikedanese** [comment]
@bgrant0607 so far kubeadm will first appear in the v1.5 release tar because the code hadn't merged so no, it probably hasn't been used much yet. We want pulling kubeadm from the release tar to be an option for people who aren't using debian or rhel.

**Comment by mikedanese** [comment]
How does per-PR testing work when we move out. Do we test core PRs against master of kubeadm and test kubeadm PRs against master of core? It seems like two competing SQs could interact (i.e. break) in ways that would halt both SQs. cc @kubernetes/sig-testing

**Comment by zmerlynn** [comment]
The way I did this for `kops` is this:
* There's a PR builder on the main `kops` repo that runs the equivalent of [`ci-kubernetes-e2e-gce`]( https://k8s-gubernator.appspot.com/builds/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce/), but pinned to a stable Kubernetes version.
* There's a [`kops-updown`](https://k8s-testgrid.appspot.com/google-aws#kops-aws-updown) job that pushes to a `kops` release file, `gs://kops-ci/bin/latest-ci-updown-green.txt`, on green.
* `latest-ci-updown-green.txt` is consumed by the [`kops`](https://k8s-testgrid.appspot.com/google-aws#kops-aws) job and pushes to `gs://kops-ci/bin/latest-ci-green.txt` on green.
* The Kubernetes repo PR builder consumes `gs://kops-ci/bin/latest-ci-green.txt` to run against Kubernetes PRs.

**Comment by pipejakob** [comment]
First draft of proposal is [here](https://docs.google.com/a/google.com/document/d/1VUzdWnsvvV6bRCrF7Azrt4xPCErNXRnuV2nrb0By48k/edit?usp=sharing). It references and borrows from the [kubectl extraction plan](https://docs.google.com/a/google.com/document/d/1ELBwwBstbNEK8906Qz5G-A9oqJWluFZIIWModYTHe6w/edit?usp=sharing).

**Comment by timothysc** [comment]
I'm actually going to close this.  

There is no intent to be 1st mover and we will wait until other tools have vetted the process.  
When the process is well defined and stable, we will move.  Until that is done, this issue doesn't need to remain open.
